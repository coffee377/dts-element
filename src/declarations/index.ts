export * from './class-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './interface-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
