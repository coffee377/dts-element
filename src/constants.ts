import * as ts from 'typescript';
import {create_native_type, INativeType} from './types/native-type';

export enum ElementKind {
  GenericDeclaration,
  GenericType,
  NativeType,
  ArrayType,
  Parameter,
}

export const any_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)});
export const never_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NeverKeyword)});

export const null_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NullKeyword)});
export const undefined_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.UndefinedKeyword)});
export const void_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword)});

export const boolean_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword)});
export const number_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)});
export const string_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)});
export const symbol_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.SymbolKeyword)});
export const object_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword)});

export const this_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.ThisKeyword)});
