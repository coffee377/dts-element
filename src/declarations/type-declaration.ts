import * as ts from 'typescript';
import {IType} from '../collections';
import {any_type} from '../constants';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {transform} from '../transform';
import {create_type_parameters} from '../utils';
import {IGenericDeclaration} from './generic-declaration';

export interface ITypeDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  type?: IType;
}

export interface ITypeDeclaration
  extends IElement<ElementKind.TypeDeclaration>, ITypeDeclarationOptions {}

export const create_type_declaration = (options: ITypeDeclarationOptions): ITypeDeclaration => ({
  ...create_element(ElementKind.TypeDeclaration),
  ...options,
});

// tslint:disable:ter-indent

export const transform_type_declaration = (element: ITypeDeclaration, path: IElement<any>[]) => {
  const type_declaration = ts.createTypeAliasDeclaration(
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* type            */ transform(element.type || any_type, path) as ts.TypeNode,
  );
  if (element.export === true) {
    type_declaration.modifiers = [ts.createToken(ts.SyntaxKind.ExportKeyword)] as ts.NodeArray<ts.Modifier>;
  }
  return type_declaration;
};
