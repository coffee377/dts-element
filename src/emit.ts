import * as ts from 'typescript';
import {ElementKind} from './constants';
import {IElement} from './element';
import {emit_triple_slash_reference, ITripleSlashReference} from './others/triple-slash-reference';
import {transform} from './transform';

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) => {
  switch (element.kind) {
    case ElementKind.TripleSlashReference:
      return emit_triple_slash_reference(element as ITripleSlashReference);
    default:
      const ts_node = transform(element);
      const source_file = ts.createSourceFile('', '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
      const printer = ts.createPrinter(options);
      return printer.printNode(ts.EmitHint.Unspecified, ts_node, source_file);
  }
};
