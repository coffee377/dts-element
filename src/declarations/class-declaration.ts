import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {IClassMember} from '../members';
import {IIndexSignature} from '../others/index-signature';
import {transform} from '../transform';
import {IGeneralType} from '../types/general-type';
import {add_declare_modifier_if_need, create_type_nodes, create_type_parameters} from '../utils';
import {IGenericDeclaration} from './generic-declaration';

export interface IClassDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  members?: (IClassMember | IIndexSignature)[];
  abstract?: boolean;
  extends?: IGeneralType;
}

export interface IClassDeclaration
  extends IElement<ElementKind.ClassDeclaration>, IClassDeclarationOptions {}

export const create_class_declaration = (options: IClassDeclarationOptions): IClassDeclaration => ({
  ...create_element(ElementKind.ClassDeclaration),
  ...options,
});

// tslint:disable:ter-indent max-line-length

export const transform_class_declaration = (element: IClassDeclaration, path: IElement<any>[]) =>
  ts.createClassDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
                            [
                              ...(element.export === true)
                                ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                                : [],
                              ...(element.abstract === true)
                                ? [ts.createToken(ts.SyntaxKind.AbstractKeyword)]
                                : [],
                            ],
                            path,
                          ),
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* heritageClauses */ (element.extends === undefined)
                            ? []
                            : [ts.createHeritageClause(
                            /* token */ ts.SyntaxKind.ExtendsKeyword,
                            /* types */ [ts.createExpressionWithTypeArguments(
                                          /* typeArguments */ create_type_nodes(element.extends.generics, path),
                                          /* expression    */ ts.createIdentifier(element.extends.name),
                                        )],
                            )],
    /* members         */ (element.members || []).map(member => transform(member, path) as ts.ClassElement),
  );
