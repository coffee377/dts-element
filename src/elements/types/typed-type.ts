import {emit_declaration_type_name} from '../../helpers/emit-declaration-type-name';
import {emit_generics} from '../../helpers/emit-generics';
import {Stack} from '../../stack';
import {TypeDeclaration} from '../declarations/type-declaration';
import {Type} from '../type';

export interface ITypedTypeRequiredParameters {
  owned: string | TypeDeclaration;
}

export interface ITypedTypeOptionalParameters {
  generics: Type[];
}

export class TypedType extends Type<ITypedTypeRequiredParameters, ITypedTypeOptionalParameters> {

  public get default_parameters(): ITypedTypeOptionalParameters {
    return {
      generics: [],
    };
  }

  public _emit(stack: Stack): string {
    const name = emit_declaration_type_name(this.parameters.owned);
    return `${name}${emit_generics(this.parameters.generics, stack)}`;
  }

}
