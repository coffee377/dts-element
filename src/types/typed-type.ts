import * as ts from 'typescript';
import {IType} from '../collections';
import {ElementKind} from '../constants';
import {INamespaceDeclaration} from '../declarations/namespace-declaration';
import {ITypeDeclaration} from '../declarations/type-declaration';
import {create_element, IElement, IElementOptions} from '../element';
import {transform_general_type} from './general-type';

export interface ITypedTypeOptions extends IElementOptions {
  parents?: (string | INamespaceDeclaration)[];
  name: string | ITypeDeclaration;
  generics?: IType[];
}

export interface ITypedType
  extends IElement<ElementKind.TypedType>, ITypedTypeOptions {}

export const create_typed_type = (options: ITypedTypeOptions): ITypedType => ({
  ...create_element(ElementKind.TypedType),
  ...options,
});

export const transform_typed_type: (element: ITypedType, path: IElement<any>[]) => ts.TypeReferenceNode
  = transform_general_type as any;
