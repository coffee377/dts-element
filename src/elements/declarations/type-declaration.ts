import {any_type} from '../../constants';
import {emit_export} from '../../helpers/emit-export';
import {emit_generics} from '../../helpers/emit-generics';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {Type} from '../type';
import {GenericType} from '../types/generic-type';

export interface ITypeDeclarationRequiredParameters {
  name: string;
}

export interface ITypeDeclarationOptionalParameters {
  generics: GenericType[];
  target: Type;
  export: boolean;
}

export class TypeDeclaration
    extends Declaration<ITypeDeclarationRequiredParameters, ITypeDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & ITypeDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      target: any_type,
      export: false,
    });
  }

  public _emit_raw(stack: Stack): string {
    const {name, generics} = this.parameters;
    const generic = emit_generics(generics, stack);
    const target = this.parameters.target.emit(stack);
    const an_export = emit_export(this.parameters.export, stack);
    return `${an_export}type ${name}${generic} = ${target};`;
  }

}
