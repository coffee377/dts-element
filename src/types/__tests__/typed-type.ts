import {any_type, string_type} from '../../constants';
import {create_type_declaration} from '../../declarations/type-declaration';
import {emit} from '../../emit';
import {create_typed_type} from '../typed-type';

it('should return correctly with name (string)', () => {
  expect(emit(
    create_typed_type({
      name: 'Something',
    }),
  )).toMatchSnapshot();
});

it('should return correctly with name (TypeDeclaration)', () => {
  expect(emit(
    create_typed_type({
      name: create_type_declaration({
        name: 'Something',
      }),
    }),
  )).toMatchSnapshot();
});

it('should return correctly with name, generics', () => {
  expect(emit(
    create_typed_type({
      name: 'Something',
      generics: [
        any_type,
        string_type,
      ],
    }),
  )).toMatchSnapshot();
});
