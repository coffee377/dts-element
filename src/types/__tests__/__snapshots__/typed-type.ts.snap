// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with name (TypeDeclaration) 1`] = `"Something"`;

exports[`should return correctly with name (string) 1`] = `"Something"`;

exports[`should return correctly with name, generics 1`] = `"Something<any, string>"`;
